### PR TITLE
Add command support

### PR DESCRIPTION
**Pull request by osamadawoood**
Closes: https://github.com/microsoft/dtdl-model-generator/issues/13

### CODE REVIEW

File: test/Generator.Tests.Generated/EnumOutputComplexCommand.cs [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+// Licensed under the MIT License.
+
+namespace Generator.Tests.Generated
+{
+    using System.ComponentModel.DataAnnotations;
+    using System.Runtime.Serialization;
+    using System.Text.Json.Serialization;
+
+    [JsonConverter(typeof(JsonStringEnumConverter))]
+    public enum EnumOutputComplexCommand
```
Comment (osamadawoood) [inline test/Generator.Tests.Generated/EnumOutputComplexCommand.cs:11]: Come to think of it this should probably be renamed to AssetComplexCommandResponseEnum. @jarz what do you think?

Comment (jarz) [reply test/Generator.Tests.Generated/EnumOutputComplexCommand.cs:11]: I _think_ it should be `AssetComplexCommandResponse` -- `POISubStatus` is another `enum` that's generated.

Comment (osamadawoood) [reply test/Generator.Tests.Generated/EnumOutputComplexCommand.cs:11]: Can an Enum be nested in an object response?

File: src/Generator/Base/Command.cs [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+                    var dateOnlyProperty = new DateOnlyProperty(commandPayloadInfo, Options);
+                    return dateOnlyProperty.Type;
+
+                default:
+                    if (!Types.TryGetNullable(commandPayloadInfo.Schema.EntityKind, out var type))
+                    {
+                        throw new Exception($"Unsupported primitive property type: {commandPayloadInfo.Schema.EntityKind} for {commandPayloadInfo.Name} in {commandName} command in {enclosingClass}");
+                    }
+
+                    return type ?? throw new Exception($"Unsupported primitive property type: {commandPayloadInfo.Schema.EntityKind} for {commandPayloadInfo.Name} in {commandName} command in {enclosingClass}");
```
[the remark this answers is not in the document]
Comment (osamadawoood) [reply src/Generator/Base/Command.cs:97]: type could be a null, no?
`dtdl-model-generator\src\Generator\Base\Command.cs(102,24): warning CS8603: Possible null reference return. [C:\Users\osa
madawood\source\repos\dtdl-model-generator\src\Generator\Generator.csproj]`

[the remark this answers is not in the document]
Comment (osamadawoood) [reply src/Generator/Base/Command.cs:97]: Nvm, this comment solves it: https://github.com/microsoft/dtdl-model-generator/pull/26#discussion_r1140536954

File: src/Generator/Entity/ModelEntity.cs [hunk 3]
```diff
@@ -16,8 +16,10 @@ internal ModelEntity(DTInterfaceInfo interfaceInfo, ModelGeneratorOptions option
         Name = GetClassName(interfaceInfo.Id);
         Parent = interfaceInfo.Extends.Count() > 0 ? GetClassName(interfaceInfo.Extends.First().Id) : nameof(BasicDigitalTwin);
         FileDirectory = ExtractDirectory(ModelId);
-        var contents = interfaceInfo.Contents.Select(c => c.Value).Where(c => c.Id.Labels.Contains(Name));
-        Content.AddRange(contents.Select(CreateProperty));
+        var contents = interfaceInfo.Contents.Select(c => c.Value).Where(c => c.Id.Labels.Contains(Name) && c is not DTCommandInfo);
+        PropertyContent.AddRange(contents.Select(CreateProperty));
+        var commands = interfaceInfo.Contents.Select(c => c.Value).Where(c => c.Id.Labels.Contains(Name) && c is DTCommandInfo);
+        CommandContent.AddRange(commands.Select((contentInfo) => new Command((DTCommandInfo)contentInfo, Name, Options)));
     }
```
Comment (jarz) [inline src/Generator/Entity/ModelEntity.cs:23]: Some thoughts...

```suggestion
        var properties = NEEDS_A_DECENT_NAME<DTPropertyInfo>(interfaceInfo);
        PropertyContent.AddRange(properties.Select(CreateProperty));
        var relationships = NEEDS_A_DECENT_NAME<DTRelationshipInfo>(interfaceInfo);
        PropertyContent.AddRange(relationships.Select(CreateProperty));
        var commands = NEEDS_A_DECENT_NAME<DTCommandInfo>(interfaceInfo);
        CommandContent.AddRange(commands.Select((contentInfo) => new Command((DTCommandInfo)contentInfo, Name, Options)));
    }

    // ... protected methods go here ...

    // This method *could* probably exist in a parent class, but we can move it "up" later on if there's a need...
    private IEnumerable<DTContentInfo> NEEDS_A_DECENT_NAME<T>(DTInterfaceInfo interfaceInfo) where T : DTContentInfo // <-- I think that's the right type constraint...
    {
        return interfaceInfo.Contents.Select(c => c.Value).Where(c => c.Id.Labels.Contains(Name) && c is T);
    }
```
Maybe `NEEDS_A_DECENT_NAME` could be `SelectContentInfo`?

I'd also consider creating a `CreateCommand()` in the `ClassEntity` class (matching the pattern used with `CreateProperty()`):
```csharp
    protected Command CreateCommand(DTContentInfo content)
    {
        if (content is DTCommandInfo command)
        {
            return new Command(command, Name, Options);
        }

        throw new Exception($"Unsupported content type: {content.EntityKind}");
    }
```
(Nilay's PR #24 introduces `UnsupportedContentTypeException` that should be used instead, but maybe we address it in that PR if this one goes first...)

File: src/Generator/Exceptions/UnsupportedPrimativeTypeException .cs [hunk 4]
```diff
@@ -0,0 +1,41 @@
+// Copyright (c) Microsoft Corporation.
```
Comment (jarz) [inline src/Generator/Exceptions/UnsupportedPrimativeTypeException .cs:1]: Extra space at the end of the filename.

**Review by jarz — CHANGES_REQUESTED** [review]

File: src/Generator.TemplateProject/Custom/CloudToDeviceMethodOptions.cs [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+    internal CloudToDeviceMethodOptions(TimeSpan? connectionTimeout, TimeSpan? responseTimeout)
+    {
+        ConnectionTimeout = connectionTimeout;
+        ResponseTimeout = responseTimeout;
+    }
+
+    internal TimeSpan? ConnectionTimeout { get; set; }
+
+    internal TimeSpan? ResponseTimeout { get; set; }
```
[the remark this answers is not in the document]
Comment (jarz) [reply src/Generator.TemplateProject/Custom/CloudToDeviceMethodOptions.cs]: @osamadawoood - this didn't address point 2

**Review by jarz — CHANGES_REQUESTED** [review]
One (fairly small) change to make -- but really important for users!

File: src/Generator.TemplateProject/Custom/CloudToDeviceMethodOptions.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (osamadawoood) [reply src/Generator.TemplateProject/Custom/CloudToDeviceMethodOptions.cs]: Ops. Updated.

**Review by nilay-kapadia — APPROVED** [review]

**Review by jarz — APPROVED** [review]

**Pull request merged**